[Bug] [connector] [ftp] Read timed out

### Search before asking

- [X] I had searched in the [issues](https://github.com/DTStack/chunjun/issues) and found no similar issues.


### What happened

local 模式1.16 运行 ftp->ftp sql 作业报错

```
Caused by: com.dtstack.chunjun.throwable.ChunJunRuntimeException: java.net.SocketTimeoutException: Read timed out
	at com.dtstack.chunjun.connector.ftp.sink.FtpOutputFormat.wrapFtpHandler(FtpOutputFormat.java:394)
	at com.dtstack.chunjun.connector.ftp.sink.FtpOutputFormat.deleteTmpDataDir(FtpOutputFormat.java:94)
	at com.dtstack.chunjun.sink.format.BaseFileOutputFormat.initializeGlobal(BaseFileOutputFormat.java:70)
	at org.apache.flink.runtime.jobgraph.InputOutputFormatVertex.initializeOnMaster(InputOutputFormatVertex.java:112)
	at org.apache.flink.runtime.executiongraph.DefaultExecutionGraphBuilder.buildGraph(DefaultExecutionGraphBuilder.java:176)
	... 15 more
Caused by: java.net.SocketTimeoutException: Read timed out
	at java.net.SocketInputStream.socketRead0(Native Method)
	at java.net.SocketInputStream.socketRead(SocketInputStream.java:116)
	at java.net.SocketInputStream.read(SocketInputStream.java:171)
	at java.net.SocketInputStream.read(SocketInputStream.java:141)
	at sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:284)
	at sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:326)
	at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:178)
	at java.io.InputStreamReader.read(InputStreamReader.java:184)
	at java.io.BufferedReader.fill(BufferedReader.java:161)
	at java.io.BufferedReader.read(BufferedReader.java:182)
	at org.apache.commons.net.io.CRLFLineReader.readLine(CRLFLineReader.java:58)
	at org.apache.commons.net.ftp.FTP.__getReply(FTP.java:310)
	at org.apache.commons.net.ftp.FTP.__getReply(FTP.java:290)
	at org.apache.commons.net.ftp.FTP.getReply(FTP.java:637)
	at org.apache.commons.net.ftp.FTPClient.completePendingCommand(FTPClient.java:1637)
	at com.dtstack.chunjun.connector.ftp.handler.FtpHandler.isFileExist(FtpHandler.java:159)
	at com.dtstack.chunjun.connector.ftp.handler.FtpHandler.deleteAllFilesInDir(FtpHandler.java:341)
	at com.dtstack.chunjun.connector.ftp.sink.FtpOutputFormat.lambda$deleteTmpDataDir$2(FtpOutputFormat.java:94)
	at com.dtstack.chunjun.connector.ftp.sink.FtpOutputFormat.wrapFtpHandler(FtpOutputFormat.java:392)
	... 19 more
```

### What you expected to happen

--

### How to reproduce



```

CREATE TABLE SOURCE (
    C_ID INT,
    C_BOOLEAN BOOLEAN,
    C_TINYINT TINYINT,
    C_SMALLINT SMALLINT,
    C_INT INT,
    C_BIGINT BIGINT,
    C_FLOAT FLOAT,
    C_DOUBLE DOUBLE,
    C_DECIMAL DECIMAL,
    C_STRING STRING,
    C_VARCHAR VARCHAR(255),
    C_CHAR CHAR(255),
    C_TIMESTAMP TIMESTAMP,
    C_DATE DATE
) WITH (
    'connector' = 'ftp-x',
    -- 路径不支持通配符*,支持传目录或者具体文件地址
    'path' = '/flinkx-test/source-zip',
    -- 服务器访问协议，目前支持ftp、sftp
    'protocol' = 'ftp',
    'host' = 'xxxx',
    -- 默认值：若传输协议是sftp协议，默认值是22；若传输协议是标准ftp协议，默认值是21
    'port' = '21',
    'username' = 'xxxx',
    'password' = 'xxx',
    -- 文件的类型，和原生flink保持一致，支持原生所有类型。 读取的文件类型，默认取文件后缀名，支持CSV,TXT,EXCEL
    'format' = 'csv',
    -- protocol为ftp时的连接模式，可选PASV和PORT
    'connect-pattern' = 'PASV',
    -- 连接超时时间，单位毫秒
    'timeout' = '5000',
    -- isFirstLineHeader 首行是否为标题行，如果是则不读取第一行 默认值：false
    'first-line-header' = 'true',
    -- 读取的文件类型，默认取文件后缀名，支持CSV,TXT,EXCEL
    'file-type' = 'TXT',
    -- 读取的字段分隔符 默认值：,
    'field-delimiter' = '\t',
    -- 读取文件的编码配置 默认值：UTF-8
    'encoding' = 'UTF-8',
    -- 文件压缩类型,支持ZIP 默认值：无
    'compress-type' = 'ZIP'
    -- 以下sql模式暂不支持的参数
    -- fileConfig 文件参数配置 csv文件是否进行trim：`"fileConfig":{"trimWhitespace":true}`
);

CREATE TABLE SINK (
    C_ID INT,
    C_BOOLEAN BOOLEAN,
    C_TINYINT TINYINT,
    C_SMALLINT SMALLINT,
    C_INT INT,
    C_BIGINT BIGINT,
    C_FLOAT FLOAT,
    C_DOUBLE DOUBLE,
    C_DECIMAL DECIMAL,
    C_STRING STRING,
    C_VARCHAR VARCHAR(255),
    C_CHAR CHAR(255),
    C_TIMESTAMP TIMESTAMP,
    C_DATE DATE
) WITH (
    'connector' = 'ftp-x',
    -- 会自动创建目录
    'path' = '/flinkx-test/sink',
    -- 服务器访问协议，目前支持ftp、sftp
    'protocol' = 'ftp',
    'host' = 'xxx',
    'port' = '21',
    'username' = 'xxx',
    'password' = 'xxx',
    -- 文件的类型，和原生flink保持一致，支持原生所有类型 默认值：csv
    'format' = 'csv',
    -- 文件名, 当文件大小超过1G时，会此名_x_x 命名文件
    'file-name' = 'flinkx-sink',
    -- 文件扩展名
    'file-type' = '.log',
    -- 写入的的字段分隔符 默认值：,
    'csv.field-delimiter' = '\t'
    -- 不支持 输出标题行
    -- 写入ftp单个文件最大大小，单位字节, 默认大小1G
    ,'max-file-size' = '10485760' -- 10M
);


INSERT INTO SINK
SELECT *
FROM SOURCE;



```

### Anything else

_No response_

### Version

master

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

--- a/chunjun-connectors/chunjun-connector-ftp/src/main/java/com/dtstack/chunjun/connector/ftp/converter/DummyInitializationContext.java
+++ b/chunjun-connectors/chunjun-connector-ftp/src/main/java/com/dtstack/chunjun/connector/ftp/converter/DummyInitializationContext.java
@@ -1,0 +1,42 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.dtstack.chunjun.connector.ftp.converter;
+
+import org.apache.flink.api.common.serialization.DeserializationSchema;
+import org.apache.flink.api.common.serialization.SerializationSchema;
+import org.apache.flink.metrics.MetricGroup;
+import org.apache.flink.metrics.groups.UnregisteredMetricsGroup;
+import org.apache.flink.util.SimpleUserCodeClassLoader;
+import org.apache.flink.util.UserCodeClassLoader;
+
+/** A dummy context for serialization schemas. */
+public class DummyInitializationContext
+        implements SerializationSchema.InitializationContext,
+                DeserializationSchema.InitializationContext {
+
+    @Override
+    public MetricGroup getMetricGroup() {
+        return new UnregisteredMetricsGroup();
+    }
+
+    @Override
+    public UserCodeClassLoader getUserCodeClassLoader() {
+        return SimpleUserCodeClassLoader.create(DummyInitializationContext.class.getClassLoader());
+    }
+}

--- a/chunjun-connectors/chunjun-connector-ftp/src/main/java/com/dtstack/chunjun/connector/ftp/converter/FtpSqlConverter.java
+++ b/chunjun-connectors/chunjun-connector-ftp/src/main/java/com/dtstack/chunjun/connector/ftp/converter/FtpSqlConverter.java
@@ -43,11 +43,13 @@ public class FtpSqlConverter extends AbstractRowConverter<String, String, String
 
     @Override
     public RowData toInternal(String input) throws Exception {
+        valueDeserialization.open(new DummyInitializationContext());
         return valueDeserialization.deserialize(input.getBytes());
     }
 
     @Override
     public String toExternal(RowData rowData, String output) throws Exception {
+        valueSerialization.open(new DummyInitializationContext());
         byte[] serialize = valueSerialization.serialize(rowData);
         return new String(serialize);
     }

--- a/chunjun-connectors/chunjun-connector-ftp/src/main/java/com/dtstack/chunjun/connector/ftp/handler/FtpHandler.java
+++ b/chunjun-connectors/chunjun-connector-ftp/src/main/java/com/dtstack/chunjun/connector/ftp/handler/FtpHandler.java
@@ -149,14 +149,19 @@ public class FtpHandler implements DTFtpHandler {
     @Override
     public boolean isFileExist(String filePath) throws IOException {
         ftpClient.enterLocalPassiveMode();
-        try (InputStream inputStream = ftpClient.retrieveFileStream(encodePath(filePath))) {
-            return inputStream != null && ftpClient.getReplyCode() != 550;
+        InputStream inputStream = null;
+        try {
+            inputStream = ftpClient.retrieveFileStream(encodePath(filePath));
+            return inputStream != null && ftpClient.getReplyCode() != FTPReply.FILE_UNAVAILABLE;
         } catch (IOException e) {
             throw new ChunJunRuntimeException(
                     "An exception occurred when judging whether the file exists. filepath: "
                             + filePath);
         } finally {
-            ftpClient.completePendingCommand();
+            if (inputStream != null) {
+                inputStream.close();
+                ftpClient.completePendingCommand();
+            }
         }
     }
 

--- a/chunjun-connectors/chunjun-connector-ftp/src/main/java/com/dtstack/chunjun/connector/ftp/sink/FtpOutputFormat.java
+++ b/chunjun-connectors/chunjun-connector-ftp/src/main/java/com/dtstack/chunjun/connector/ftp/sink/FtpOutputFormat.java
@@ -49,6 +49,8 @@ public class FtpOutputFormat extends BaseFileOutputFormat {
     /** 换行符 */
     private static final int NEWLINE = 10;
 
+    private static final String SP = "/";
+
     private static final long serialVersionUID = -3331910729974811530L;
 
     protected FtpConfig ftpConfig;
@@ -101,7 +103,7 @@ public class FtpOutputFormat extends BaseFileOutputFormat {
         if (writer != null) {
             return;
         }
-        String currentBlockTmpPath = tmpPath + File.separatorChar + currentFileName;
+        String currentBlockTmpPath = tmpPath + SP + currentFileName;
         try {
             os = ftpHandler.getOutputStream(currentBlockTmpPath);
             writer =
@@ -212,7 +214,7 @@ public class FtpOutputFormat extends BaseFileOutputFormat {
                 String fileName =
                         handleUserSpecificFileName(
                                 tmpDataFile.getName(), dataFiles.size(), copyList, ftpHandler);
-                String newFilePath = outputFilePath + File.separatorChar + fileName;
+                String newFilePath = outputFilePath + SP + fileName;
                 ftpHandler.rename(currentFilePath, newFilePath);
                 copyList.add(newFilePath);
                 log.info("copy temp file:{} to dir:{}", currentFilePath, outputFilePath);
@@ -253,7 +255,7 @@ public class FtpOutputFormat extends BaseFileOutputFormat {
                         List<String> copyList = new ArrayList<>();
                         for (String dataFile : dataFiles) {
                             File tmpDataFile = new File(dataFile);
-                            dataFilePath = tmpDataFile.getAbsolutePath();
+                            dataFilePath = dataFile;
 
                             String fileName =
                                     handleUserSpecificFileName(
@@ -261,8 +263,7 @@ public class FtpOutputFormat extends BaseFileOutputFormat {
                                             dataFiles.size(),
                                             copyList,
                                             handler);
-                            String newDataFilePath =
-                                    ftpConfig.getPath() + File.separatorChar + fileName;
+                            String newDataFilePath = ftpConfig.getPath() + SP + fileName;
                             handler.rename(dataFilePath, newDataFilePath);
                             copyList.add(newDataFilePath);
                             log.info("move temp file:{} to dir:{}", dataFilePath, outputFilePath);
@@ -285,12 +286,15 @@ public class FtpOutputFormat extends BaseFileOutputFormat {
 
     @Override
     protected String getExtension() {
+        if (StringUtils.isNotBlank(ftpConfig.getFileType())) {
+            return ftpConfig.getFileType();
+        }
         return "";
     }
 
     @Override
     protected long getCurrentFileSize() {
-        String path = tmpPath + File.separatorChar + currentFileName;
+        String path = tmpPath + SP + currentFileName;
         try {
             return ftpHandler.getFileSize(path);
         } catch (IOException e) {
@@ -351,7 +355,7 @@ public class FtpOutputFormat extends BaseFileOutputFormat {
     private String filepathCheck(String filename, List<String> copyList, DTFtpHandler handler)
             throws IOException {
         if (WriteMode.NONCONFLICT.name().equalsIgnoreCase(ftpConfig.getWriteMode())) {
-            if (handler.isFileExist(ftpConfig.getPath() + File.separatorChar + filename)) {
+            if (handler.isFileExist(ftpConfig.getPath() + SP + filename)) {
                 handler.deleteAllFilesInDir(tmpPath, null);
                 if (!copyList.isEmpty()) {
                     for (String filePath : copyList) {
@@ -367,7 +371,7 @@ public class FtpOutputFormat extends BaseFileOutputFormat {
                         String.format("the file: %s already exists", filename));
             }
         } else if (WriteMode.INSERT.name().equalsIgnoreCase(ftpConfig.getWriteMode())) {
-            if (handler.isFileExist(ftpConfig.getPath() + File.separatorChar + filename)) {
+            if (handler.isFileExist(ftpConfig.getPath() + SP + filename)) {
                 String suffix =
                         StringUtils.isNotBlank(ftpConfig.getSuffix())
                                 ? "_" + ftpConfig.getSuffix()

--- a/chunjun-connectors/chunjun-connector-ftp/src/main/java/com/dtstack/chunjun/connector/ftp/spliter/ConcurrentZipCompressSplit.java
+++ b/chunjun-connectors/chunjun-connector-ftp/src/main/java/com/dtstack/chunjun/connector/ftp/spliter/ConcurrentZipCompressSplit.java
@@ -54,7 +54,7 @@ public class ConcurrentZipCompressSplit extends DefaultFileSplit {
             while ((zipEntry = zipInputStream.getNextEntry()) != null) {
                 fileSplits.add(
                         new FtpFileSplit(
-                                0, zipEntry.getSize(), filePath, zipEntry.getName(), compressType));
+                                0, zipEntry.getSize(), zipEntry.getName(), filePath, compressType));
             }
             return fileSplits;
         } catch (Exception e) {

--- a/chunjun-connectors/chunjun-connector-ftp/src/main/java/com/dtstack/chunjun/connector/ftp/table/FtpDynamicTableFactory.java
+++ b/chunjun-connectors/chunjun-connector-ftp/src/main/java/com/dtstack/chunjun/connector/ftp/table/FtpDynamicTableFactory.java
@@ -68,6 +68,7 @@ public class FtpDynamicTableFactory implements DynamicTableSourceFactory, Dynami
         ftpConfig.setMaxFileSize(config.get(BaseFileOptions.MAX_FILE_SIZE));
         ftpConfig.setCompressType(config.get(FtpOptions.COMPRESS_TYPE));
         ftpConfig.setFileType(config.get(FtpOptions.FILE_TYPE));
+        ftpConfig.setNextCheckRows(config.get(BaseFileOptions.NEXT_CHECK_ROWS));
 
         if (!ConfigConstants.DEFAULT_FIELD_DELIMITER.equals(
                 config.get(FtpOptions.FIELD_DELIMITER))) {
@@ -179,6 +180,7 @@ public class FtpDynamicTableFactory implements DynamicTableSourceFactory, Dynami
         options.add(FtpOptions.FIRST_LINE_HEADER);
         options.add(FtpOptions.FIELD_DELIMITER);
         options.add(FtpOptions.COMPRESS_TYPE);
+        options.add(BaseFileOptions.NEXT_CHECK_ROWS);
         return options;
     }
 }

